### PR TITLE
Memoize patched calls and support UUID conversion

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

:boom: Breaking change - now `workflow.patched()` calls always return the same boolean for the life of the workflow code. This will break users with running workflows that make multiple `workflow.patched` for the same key in the same workflow (e.g. in a loop, etc).

## Why?

We were not memoizing before and people want to be able to check the same patch key at different parts of workflow code. IT was deemed acceptable to make this break while Python was in beta.

## Checklist

1. Relates to #143

### CODE REVIEW

File: temporalio/worker/_workflow_instance.py [hunk 1]
```diff
@@ -739,7 +740,14 @@ def workflow_memo_value(
...
+        # We use a previous memoized result of this if present. If this is being
+        # deprecated, we can still use memoized result and skip the command.
+        use_patch = self._patches_memoized.get(id)
+        if use_patch is not None:
+            return use_patch
+
         use_patch = not self._is_replaying or id in self._patches_notified
+        self._patches_memoized[id] = use_patch
         # Only add patch command if never sent before for this ID
         if use_patch and not id in self._patches_sent:
```
Comment (bergundy) [inline temporalio/worker/_workflow_instance.py:752]: Don't need patches sent anymore

**Review by bergundy — APPROVED** [review]

File: temporalio/worker/_workflow_instance.py (hunk 1 above)
Comment (bergundy) [reply temporalio/worker/_workflow_instance.py:752]: Not sure what you mean, once memoized, we won't enter this path anymore so we don't need to track the extra state.

Comment (cretz) [reply temporalio/worker/_workflow_instance.py:752]: True, fixing

Comment (cretz) [reply temporalio/worker/_workflow_instance.py]: Fixed

**Comment by bergundy** [comment]
@cretz before we release this, let's make sure that we're also upserting the search attributes to avoid another history incompatible change

**Comment by cretz** [comment]
:+1: I have opened https://github.com/temporalio/sdk-python/issues/194. Can you please comment there on what you think the implementation should be?

**Pull request merged**